src/app/utils/query-url-normalization.ts

```diff
@@ -0,0 +1,136 @@
+// Matches patterns like "('<1f7ff346-c174-45e5-af38-294e51d9969a>')" or "('<key>')"
+const keyIdRegex = /\(\'\<?\{?[ ?0-9a-zA-Z-]*\}?\>?\'\)/g;
+
+// Matches patterns like "(query=<key>)" or "(itemat=<key>,mode=<mode>)"
+const functionParamInitialRegex = /(?<=\=).*?(?=,)/g;
+const functionParamFinalRegex = /(?<=\=).[^,]*(?=\))/g;
+
+// Matches pattterns within quotes e.g "displayName: Gupta"
+const quotedTextRegex = /"([^"]*)"/g;
+
+// Matches PII patterns
+const numberRegex =  /(?<=\/)\d+\b/g; // number between forward slashes
+const emailRegex = /([a-z0-9_\-.+]+)@\w+(\.\w+)*/gi;
+const guidRegex =
+ /(\{){0,1}[0-9a-fA-F]{8}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{12}(\}){0,1}/gi;
+
+
+export function normalizeQueryUrl(incomingUrl: string) {
+  incomingUrl = decodeURIComponent(incomingUrl);
+
+  // Extract query string
+  let queryString: string = '';
+  const urlParts = incomingUrl.split('?');
+  if (urlParts.length > 1) {
+    const startOfQueryString = incomingUrl.indexOf('?');
+    queryString = incomingUrl.substring(startOfQueryString, incomingUrl.length - startOfQueryString);
+    queryString = processQueryParams(queryString.slice(1));
+  }
+
+  // Drop query string
+  incomingUrl = urlParts[0];
+
+  // Drop casts that use the /$/ pattern
+  incomingUrl = incomingUrl.split('/$/')[0];
+
+  // Normalize parameters in param=<arbitraryKey> format with param={value} format
+  incomingUrl = incomingUrl.replace(functionParamInitialRegex, '<value>');
+  incomingUrl = incomingUrl.replace(functionParamFinalRegex, '<value>');
+
+  // Replace IDs and ID placeholders with generic {id}
+  incomingUrl = incomingUrl.replace(keyIdRegex, '/<id>');
+  incomingUrl = incomingUrl.replace(guidRegex, '<id>');
+  incomingUrl = incomingUrl.replace(numberRegex, '<id>');
+
+  // Redact PII
+  incomingUrl = incomingUrl.replace(emailRegex, '<email>');
+
+  // Drop entity/action namespace
+  incomingUrl = incomingUrl.replace('microsoft.graph.', '');
+
+  // Trim off $value
+  if (incomingUrl.endsWith('/$value') || incomingUrl.endsWith('/$count'))
+  {
+    incomingUrl = incomingUrl.substring(0, incomingUrl.length - 7);
+  }
+
+  // Trim off $ref
+  if (incomingUrl.endsWith('/$ref'))
+  {
+    incomingUrl = incomingUrl.substring(0, incomingUrl.length - 5);
+  }
+
+  // Trim any delta queries
+  if (incomingUrl.endsWith('/delta'))
+  {
+    incomingUrl = incomingUrl.substring(0, incomingUrl.length - 6);
+  }
+
+  return { incomingUrl, queryString };
+}
+
+function processQueryParams(queryString: string) {
+   const params = queryString.split('&');
+   let result: string = '';
+   if (params.length)
+   {
+      params.forEach(param => {
+       result += removeSensitiveQueryParamValues(param) + '&';
+      });
+      result = result.slice(0, -1);
```

Comment:
It would be helpful to add a comment to show that we are removing the last ampersand

Reply:
I've removed this part. I am using an existing function to do this.